lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -1038,6 +1048,65 @@ data ApiAccountKey = ApiAccountKey
     } deriving (Eq, Generic, Show)
       deriving anyclass NFData
 
+data ApiSharedWalletPostDataFromMnemonics = ApiSharedWalletPostDataFromMnemonics
+    { name :: !(ApiT WalletName)
+    , mnemonicSentence :: !(ApiMnemonicT (AllowedMnemonics 'Shelley))
+    , mnemonicSecondFactor :: !(Maybe (ApiMnemonicT (AllowedMnemonics 'SndFactor)))
+    , passphrase :: !(ApiT (Passphrase "raw"))
+    , accountIndex :: !(ApiT DerivationIndex)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    } deriving (Eq, Generic, Show)
+
+data ApiSharedWalletPostDataFromAccountPubX = ApiSharedWalletPostDataFromAccountPubX
+    { name :: !(ApiT WalletName)
+    , accountPublicKey :: !ApiAccountPublicKey
+    , accountIndex :: !(ApiT DerivationIndex)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    } deriving (Eq, Generic, Show)
+
+newtype ApiSharedWalletPostData = ApiSharedWalletPostData
+    { wallet :: Either ApiSharedWalletPostDataFromMnemonics ApiSharedWalletPostDataFromAccountPubX
+    } deriving (Eq, Generic, Show)
```

Comment:
The common fields can be split out.

```suggestion
data ApiSharedWalletPostDataFromMnemonic = ApiSharedWalletPostDataFromMnemonic
    { mnemonicSentence :: !(ApiMnemonicT (AllowedMnemonics 'Shelley))
    , mnemonicSecondFactor :: !(Maybe (ApiMnemonicT (AllowedMnemonics 'SndFactor)))
    , passphrase :: !(ApiT (Passphrase "raw"))
    } deriving (Eq, Generic, Show)

newtype ApiSharedWalletPostDataFromAccountXPub = ApiSharedWalletPostDataFromAccountXPub
    { accountPublicKey :: ApiAccountPublicKey
    } deriving (Eq, Generic, Show)

data ApiSharedWalletPostData = ApiSharedWalletPostData
    { name :: !(ApiT WalletName)
    , accountIndex :: !(ApiT DerivationIndex)
    , paymentScriptTemplate :: !ScriptTemplate
    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
    , walletKey :: !(Either ApiSharedWalletPostDataFromMnemonic ApiSharedWalletPostDataFromAccountXPub)
    } deriving (Eq, Generic, Show)
```

Is the address pool gap supposed to be in there as well?

Reply:
Done already in @rvl commit. So the idea behind not including pool gap in `ApiSharedWalletPostData` was to avoid situation that different cosigners use different pool gaps that could lead in inconsistencies when discovering shared addresses. So the pool gap will be set (for the time being) to one value (20 for example), and when requested sometime in the future, then will be add it. It was discussed with @KtorZ  in private communication (I had this exposed in the beginning)